### ISSUE TITLE
How to remove the jaxb jars from TeamCity REST API zip

### ISSUE THREAD

**Issue by netwolfuk** [tcplugins/tcWebHooks#43]
Until [this bug](https://youtrack.jetbrains.com/issue/TW-44654) is fixed, the jaxb libraries in the rest-api.zip (bundled with TeamCity) conflict with the ones pulled in by jersey-json (provided by teamcity) which the tcWebHooks REST API code depends on. 

This prevents the tcWebHooks REST API from persisting the webhook-templates.xml config to disk because it uses JaxB to serialise to XML.

Therefore, the `jaxb*` jars need to be removed from the `rest-api.zip`. This can be achieved with the following command.

```
cd TeamCity && \
zip -d webapps/ROOT/WEB-INF/plugins/rest-api.zip \
  server/jaxb-api-2.2.5.jar server/jaxb-impl-2.2.5.jar && \
rm webapps/ROOT/WEB-INF/plugins/.unpacked/rest-api/server/jaxb*.jar && \
bin/teamcity-server.sh stop && sleep 5 && bin/teamcity-server.sh start
```

**Comment by netwolfuk** [comment]
Until those jars are removed, the error in the logs looks like this:

`
DEBUG -   jetbrains.buildServer.SERVER - ClassCastException: attempting to cast jar:file:/opt/TeamCity-10.0/TeamCity/webapps/ROOT/WEB-INF/plugins/.unpacked/rest-api/server/jaxb-api-2.2.5.jar!/javax/xml/bind/JAXBContext.class to jar:file:/opt/Java/jdk1.8.0_65/jre/lib/rt.jar!/javax/xml/bind/JAXBContext.class.  Please make sure that you are specifying the proper ClassLoader. 
javax.xml.bind.JAXBException: ClassCastException: attempting to cast jar:file:/opt/TeamCity-10.0/TeamCity/webapps/ROOT/WEB-INF/plugins/.unpacked/rest-api/server/jaxb-api-2.2.5.jar!/javax/xml/bind/JAXBContext.class to jar:file:/opt/Java/jdk1.8.0_65/jre/lib/rt.jar!/javax/xml/bind/JAXBContext.class.  Please make sure that you are specifying the proper ClassLoader.
`

**Comment by netwolfuk** [comment]
Also, XML serialisation of the REST API data does not work - although "Accept:application/json" does work fine. The error returned in the browser is:
```
Error has occurred during request processing, status code: 500 (Internal Server Error).
Details: javax.ws.rs.WebApplicationException: com.sun.jersey.api.MessageException: A message body writer for Java class webhook.teamcity.server.rest.model.template.Templates, and Java type class webhook.teamcity.server.rest.model.template.Templates, and MIME media type application/xml was not found.

Not supported request. Check that URL, HTTP method and transferred data are correct.
```

**Comment by netwolfuk** [comment]
On the TeamCity Docker images, one needs to install zip first.
`apt-get install zip`

**Comment by netwolfuk** [comment]
Tool for removign offending jars is bundled with v1.1-alpha11.180.231

**Comment by Vadim-Zenin** [comment]
Could the removing offending jars tool works with TeamCity server deployed with docker?

**Comment by netwolfuk** [comment]
As you have noticed, the jars return if the docker container is stopped and started.

It might be possible to snapshot the container before stopping. https://stackoverflow.com/a/22379305

Or create volumes for the two locations listed on the removal page.

**Comment by jshumaker** [comment]
I just had to deal with this in the docker container, mainly i had applied fix but then adjusted container.  I moved to building my own image off the official ones, for example this Dockerfile will build a docker image with the offending pieces removed:

```
FROM jetbrains/teamcity-server:2019.1.3

# Apply tcWebHooks workaround, this can be removed once this issue is fixed (currently targeted for 2019.1.4!):
# https://youtrack.jetbrains.com/issue/TW-57311
RUN apt-get update && \
    apt-get install -y --no-install-recommends zip && \
    rm -rf /var/lib/apt/lists/* && \
    cd /opt/teamcity && \
    zip -d webapps/ROOT/WEB-INF/plugins/rest-api.zip server/jaxb-api-2.2.5.jar server/jaxb-impl-2.2.5.jar
```

Note it doesn't remove from .unpacked as that doesn't exist within the image, only once a container has been brought up.

I further combine this with docker-compose to simplify the process of applying updates.  It seems like this might finally get fixed with 2019.1.4 though.

**Comment by netwolfuk** [comment]
Thanks @jshumaker. This is a really nice solution. Thanks for posting.
I'm very much hoping the jaxb jars will be removed in an upcoming TeamCity release, and this issue will be resolved.

**Comment by netwolfuk** [comment]
BTW @jshumaker  that youtrack 57311 gets moved to the next release every time there is a new version of TeamCity. It's been imminent for a few years.

**Comment by jshumaker** [comment]
Sigh I was fearing that but didn't take a look at the history.  Pretty familiar with things slipping like that from Jetbrains. I've been waiting for non-horrible clean up rules for 7+ years.  I guess this workaround will live in my Dockerfile for a long time more likely.

**Comment by netwolfuk** [comment]
I'm wondering if I push your config to docker hub and run a build to build new images whenever the TeamCity images are updated

**Comment by netwolfuk** [comment]
Also, I am making progress on #122 although I've not pushed anything yet.

**Comment by netwolfuk** [comment]
FYI, @jshumaker  It looks like this will be resolved in 2019.2
See https://youtrack.jetbrains.com/issue/TW-57311

Hopefully after that, we won't have to worry about the docker dance.
